test/unit/org/apache/cassandra/db/streaming/CassandraEntireSSTableStreamWriterTest.java

```diff
@@ -126,43 +152,98 @@ public void testBlockWriterOverWire() throws IOException
-        EmbeddedChannel channel = createMockNettyChannel(serializedFile);
-        AsyncStreamingOutputPlus out = new AsyncStreamingOutputPlus(channel);
+        InetAddressAndPort peer = FBUtilities.getBroadcastAddressAndPort();
+        StreamSession session = setupStreamingSessionForTest();
+        ComponentManifest manifest = CassandraOutgoingFile.getComponentManifest(sstable);
+        CassandraStreamHeader header = header(manifest);
 
-        writer.write(out);
+        Future<?> future = sstable.runWithReadLock(ignored -> {
+            CassandraEntireSSTableStreamWriter writer = new CassandraEntireSSTableStreamWriter(sstable, session, manifest);
+
+            // start concurrent sstable component mutation
+            Future<?> mutation = CompletableFuture.completedFuture(null);
+            if (componentMutation != null)
+            {
+                mutation = ScheduledExecutors.optionalTasks.submit(componentMutation);
+                // wait 1s for concurrent task to start
+                FBUtilities.sleepQuietly(1000);
+            }
+
+            EmbeddedChannel channel = createMockNettyChannel(serializedFile);
+            AsyncStreamingOutputPlus out = new AsyncStreamingOutputPlus(channel);
+
+            writer.write(out);
+            return mutation;
+        });
+
+        // wait for concurrent component mutation to finish
+        future.get(15, TimeUnit.SECONDS);
 
         session.prepareReceiving(new StreamSummary(sstable.metadata().id, 1, 5104));
 
-        CassandraStreamHeader header =
-            CassandraStreamHeader.builder()
-                                 .withSSTableFormat(sstable.descriptor.formatType)
-                                 .withSSTableVersion(sstable.descriptor.version)
-                                 .withSSTableLevel(0)
-                                 .withEstimatedKeys(sstable.estimatedKeys())
-                                 .withSections(Collections.emptyList())
-                                 .withSerializationHeader(sstable.header.toComponent())
-                                 .withComponentManifest(CassandraOutgoingFile.getComponentManifest(sstable))
-                                 .isEntireSSTable(true)
-                                 .withFirstKey(sstable.first)
-                                 .withTableId(sstable.metadata().id)
-                                 .build();
-
-        CassandraEntireSSTableStreamReader reader = new CassandraEntireSSTableStreamReader(new StreamMessageHeader(sstable.metadata().id, peer, session.planId(), false, 0, 0, 0, null), header, session);
+        StreamMessageHeader messageHeader = new StreamMessageHeader(sstable.metadata().id, peer, session.planId(), false, 0, 0, 0, null);
+        CassandraEntireSSTableStreamReader reader = new CassandraEntireSSTableStreamReader(messageHeader, header, session);
 
         SSTableMultiWriter sstableWriter = reader.read(new DataInputBuffer(serializedFile.nioBuffer(), false));
```

Comment:
nit: Should we instantiate `sstableWriter` at the start of a try-with-resources block?

Reply:
good idea